src/tir/ir/specialize.cc

```diff
@@ -0,0 +1,326 @@
+
+    auto n = CopyOnWrite(op);
+    n->buffer = it->second;
+    n->value = std::move(value);
+    n->indices = std::move(indices);
+    return Stmt(n);
+  }
+
+  PrimExpr VisitExpr_(const BufferLoadNode* op) final {
+    auto it = buffer_map_.find(op->buffer);
+    if (it == buffer_map_.end()) {
+      return GetRef<BufferLoad>(op);
+    }
+
+    Array<PrimExpr> indices =
+        MutateArray(op->indices, [this](const PrimExpr& e) { return this->VisitExpr(e); });
+
+    auto n = CopyOnWrite(op);
+    n->buffer = it->second;
+    n->indices = std::move(indices);
+    return PrimExpr(n);
+  }
+
+  PrimExpr VisitExpr_(const VarNode* op) final {
+    auto it = var_map_.find(GetRef<Var>(op));
+    if (it == var_map_.end()) {
+      return GetRef<PrimExpr>(op);
+    } else {
+      return it->second;
+    }
+  }
+
+ private:
+  Buffer MutateBuffer(Buffer buffer) const {
+    BufferNode* buffer_ptr = buffer.CopyOnWrite();
+    Array<PrimExpr> new_shape, new_stride;
+    new_shape.reserve(buffer_ptr->shape.size());
+    new_shape.reserve(buffer_ptr->strides.size());
+    for (const auto& dim : buffer_ptr->shape) {
+      new_shape.push_back(Substitute(dim, var_map_));
+    }
+    for (const auto& stride : buffer_ptr->strides) {
+      new_shape.push_back(Substitute(stride, var_map_));
+    }
+    buffer_ptr->elem_offset = Substitute(buffer_ptr->elem_offset, var_map_);
+    buffer_ptr->shape = std::move(new_shape);
+    buffer_ptr->strides = std::move(new_stride);
+    return buffer;
+  }
+
+  Range MutateRange(const Range& range) {
+    PrimExpr min = this->VisitExpr(range->min);
+    PrimExpr extent = this->VisitExpr(range->extent);
+    if (min.same_as(range->min) && extent.same_as(range->extent)) {
+      return range;
+    } else {
+      ObjectPtr<RangeNode> n = CopyOnWrite(range.get());
+      n->min = std::move(min);
+      n->extent = std::move(extent);
+      return Range(n);
+    }
+  }
+
+  IterVar MutateIterVar(const IterVar& iter_var) {
+    Range range = MutateRange(iter_var->dom);
+    if (range.same_as(iter_var->dom)) {
+      return iter_var;
+    } else {
+      auto n = CopyOnWrite(iter_var.get());
+      n->dom = std::move(range);
+      return IterVar(n);
+    }
+  }
+
+  Buffer MutateAllocBuffer(const Buffer& alloc_buf) {
+    Buffer buf = MutateBuffer(alloc_buf);
+    if (buf.same_as(alloc_buf)) {
+      return alloc_buf;
+    } else {
+      buffer_map_[alloc_buf] = buf;
+      return buf;
+    }
+  }
+
+  BufferRegion MutateBufferRegion(const BufferRegion& buffer_region) {
+    auto it = buffer_map_.find(buffer_region->buffer);
+    Array<Range> region =
+        MutateArray(buffer_region->region,
+                    std::bind(&PrimFuncSpecializer::MutateRange, this, std::placeholders::_1));
+    if (it == buffer_map_.end() && region.same_as(buffer_region->region)) {
+      return buffer_region;
+    } else {
+      auto n = CopyOnWrite(buffer_region.get());
```

Comment:
avoid CopyOnWrite except for stmt for now, since we are not tracking cow for other things